types/index.d.ts

```diff
@@ -0,0 +1,68 @@
+import type {
```

Comment:
`microbundle` will provide types itself from JSDoc -- we should probably use that instead of a separate types file that we need to keep in sync.

Reply:
Oh, didn't know that - thanks! I will refactor the PR to use JSDoc for type definitions.

Reply:
I wasn't able to get `microbundle` to spit out types for the JSDoc on its own (also couldn't find any hints in the doc that this is supported, only [by using `.ts` files directly](https://github.com/developit/microbundle?tab=readme-ov-file#using-with-typescript)). I've tried the following:

- Added TypeScript config with `allowJs` and `emitDeclarationOnly` settings
- Added `dist/preact-custom-element.d.ts` as `"types"` in `package.json`
- Tried with and without the flag `--generateTypes=true` to enforce

Generating the types via `tsc` works as expected, but not via `microbundle` 😞 
Do you have any inputs or "hidden tricks" how to get it working @rschristian? 

_Note: I've also just seen that preact [ships the index.d.ts manually](https://github.com/preactjs/preact/blob/main/src/index.d.ts)._

Reply:
Microbundle provides TS output if the source entry is a TS file or if `"types"` is set in the `package.json` file. You might need to bump Microbundle, not sure.

> Note: I've also just seen that preact [ships the index.d.ts manually](https://github.com/preactjs/preact/blob/main/src/index.d.ts).

Preact requires hand-crafted types, it's a headache. We wouldn't do it if we didn't need to 😅

Reply:
I can try that :) If I bump microbundle it will probably also bump much more internal deps in the lockfile. What would be your preferred way of bumping?

Reply:
If bumping microbundle does work, then altering the lock file is fine, don't worry about it.

That being said, are you using an older version of `npm`? With `10.2.0`, you not only shouldn't see any changes to the lock file upon running `npm install`, but altering dep version in the `package.json` should only affect the corresponding deps in the lock file. You might have something odd in your config?

Reply:
I was using v9 as I was still on Node 18.16.1 (which wasn't compatible with NPM v10) - but I just updated :) Thanks for the hint!

Reply:
I'm not quite sure it's reliant upon v10, I think this behavior goes back to v6 or so? It's just quite odd that it seems to ignore the lock file altogether for your installs.